### ISSUE TITLE
Regression: Decorator with optional arguments gives overload error

### ISSUE THREAD

**Issue by GugelRobin** [microsoft/pyright#5384]
**Describe the bug**
See code sample.
This is most likely somehow related to the recent changes that ensured that ParamSpecs are kept unbound if used in callable that's returned. 

**To Reproduce**
Pyright 1.1.315 & 1.1.314
Lower versions don't raise an error here.
```python
from typing import overload, ParamSpec, TypeVar, Callable

T = TypeVar("T")
P = ParamSpec("P")

@overload
def foo(f: Callable[P, T]) -> Callable[P, T]:
    ...

@overload
def foo() -> Callable[[Callable[P, T]], Callable[P, T]]:
    ...

def foo(f: Callable[P, T] | None = None) -> Callable[P, T] | Callable[[Callable[P, T]], Callable[P, T]]:
    ...

# error: Overloaded implementation is not consistent with signature of overload 2
#     Function return type "((**P@foo) -> T@foo) -> ((**P@foo) -> T@foo)" is incompatible with type "((**P@foo) -> T@foo) | ((((**P@foo) -> T@foo)) -> ((**P@foo) -> T@foo))"
#       Type "((**P@foo) -> T@foo) -> ((**P@foo) -> T@foo)" cannot be assigned to type "((**P@foo) -> T@foo) | ((((**P@foo) -> T@foo)) -> ((**P@foo) -> T@foo))"
#         Type "((**P@foo) -> T@foo) -> ((**P@foo) -> T@foo)" cannot be assigned to type "(**P@foo) -> T@foo"
#         Type "((**P@foo) -> T@foo) -> ((**P@foo) -> T@foo)" cannot be assigned to type "((**P@foo) -> T@foo) -> ((**P@foo) -> T@foo)"
#           Parameter 1: type "(**P@foo) -> T@foo" cannot be assigned to type "(**P@foo) -> T@foo"
#             Type "(**P@foo) -> T@foo" cannot be assigned to type "(**P@foo) -> T@foo"
#           Function return type "(**P@foo) -> T@foo" is incompatible with type "(**P@foo) -> T@foo"
#             Type "(**P@foo) -> T@foo" cannot be assigned to type "(**P@foo) -> T@foo" (reportGeneralTypeIssues)
```

**Expected behavior**
Expected code to pass type-check.

**VS Code extension or command-line**
Running pyright 1.1.316

**Additional context**
Adding additional callable parameters that use `P` and/or `T` sometimes causes the problem not to occur - sometimes it doesn't though.

**Comment by erictraut** [comment]
Thanks for the bug report. I can confirm that I'm able to repro the problem.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.317, which I just published. It will also be included in future releases of pylance.

**Comment by karolzlot** [comment]
@erictraut 

With Pylance v2023.8.21 I am still getting an error on this code:

```python
from typing import overload

@overload
def my_function(a: int, b: int, c: int) -> int:
    ...

@overload
def my_function(a: int, c: int) -> int:
    ...

def my_function(a: int, b: int = 0, c: int = 0) -> int:
    return a + b + c
```
![image](https://github.com/microsoft/pyright/assets/11590960/6c173d8f-8473-4eb6-b400-ea3ea4b6df3c)

**Comment by erictraut** [comment]
@karolzlot, that's a different situation from the one reported in this bug. Pyright is correct to emit an error for your code sample. The implementation of your function is not compatible with the second overload because there's no way for a caller to specify `c` as a positional argument without it being interpreted as the value for parameter `b`. Not surprisingly, mypy generates the same error here as pyright.

If your intent is for `b` and `c` to be keyword-only arguments, you can fix the error by using the `*` keyword-only separator.

```python
from typing import overload

@overload
def my_function(a: int, *, b: int, c: int) -> int:
    ...

@overload
def my_function(a: int, *, c: int) -> int:
    ...

def my_function(a: int, *, b: int = 0, c: int = 0) -> int:
    return a + b + c
```
